documentation/src/main/asciidoc/userguide/chapters/query/hql/QueryLanguage.adoc

```diff
@@ -1910,7 +1997,16 @@ include::{sourcedir}/HQLTest.java[tags=hql-order-by-example]
 ====
 
 [[hql-limit-offset]]
-=== The `limit` and `offset` clauses
+=== Limits and offsets
+
+It's often useful to place a hard upper limit on the number of results that may be returned by a query.
+
+[NOTE]
+====
+_Limiting_ certainly _isn't_ a well-defined relational operation, and must be used with care.
+
+A limit can easily break the semantics of certain other features of HQL, including <<hql-explicit-fetch-join,fetch joins>>.
```

Comment:
In fact it is the other way around, fetch joins cause limits to be applied in memory 😆

Reply:
Sorry? Are you saying we now _don't_ pass through the `limit` clause when we have a fetch join, materialize the whole result set and apply the limit in Java!?

Reply:
I tested it. What's described in the documentation is the true behavior.

Reply:
Take a look into `org.hibernate.query.sqm.internal.QuerySqmImpl#doList` which throws an exception if configured, or emits a warning. If that doesn't happen, there is a bug somewhere.

Reply:
You can try it for yourself.

https://github.com/hibernate/hibernate-orm/pull/4568/files

All those queries result in a limit clause being sent to the database.

Reply:
Right, because you are fetch joining a to-one which preserves the cardinality. Try fetch joining a collection.

Reply:
Oh gawd I am too. Sorry.

Reply:
Well, changing it to a to-many collection made to difference. I still got the limit clause, and verified that I can get an incomplete collection returned.

So if you tell me that's a bug, then it's a bug.

Honestly I think both behaviors are broken, just in different ways.

Reply:
I'll look into your other PR then and try to figure out what goes wrong here. There should be an exception.